kiwi/builder/disk.py

```diff
@@ -349,26 +349,66 @@ def create_disk(self) -> Result:
+                                device_map = self._map_integrity(
+                                    device_map, integrity_root
+                                )
+                                if self.luks is not None:
+                                    with self._luks_instance(
+                                        device_map
+                                    ) as luks_root:
+                                        # create luks on raid+integrity root dev
+                                        device_map = self._map_luks(
+                                            device_map, luks_root
+                                        )
+                                        # build bootable raid+integrity+luks disk
+                                        self._process_build(device_map, disk)
+                                else:
+                                    # build bootable raid+integrity disk
+                                    self._process_build(device_map, disk)
+                        elif self.luks is not None:
+                            with self._luks_instance(device_map) as luks_root:
+                                # create luks on raid root dev
+                                device_map = self._map_luks(device_map, luks_root)
+                                # build bootable raid+luks disk
+                                self._process_build(device_map, disk)
+                        else:
+                            # build bootable raid disk
+                            self._process_build(device_map, disk)
+                elif self.integrity:
+                    # create integrity on root dev
+                    with self._integrity_instance(device_map) as integrity_root:
+                        device_map = self._map_integrity(
+                            device_map, integrity_root
+                        )
+                        if self.luks is not None:
+                            with self._luks_instance(device_map) as luks_root:
+                                # create luks on integrity root dev
+                                device_map = self._map_luks(device_map, luks_root)
+                                # build bootable integrity+luks disk
+                                self._process_build(device_map, disk)
+                        else:
+                            # build bootable integrity disk
+                            self._process_build(device_map, disk)
+                elif self.luks is not None:
+                    with self._luks_instance(device_map) as luks_root:
+                        # create luks on root dev
+                        device_map = self._map_luks(device_map, luks_root)
+                        # build bootable luks disk
+                        self._process_build(device_map, disk)
+                else:
+                    # build bootable disk
+                    self._process_build(device_map, disk)
```

Comment:
This part looks rather convoluted and a bit too deeply nested for my taste. Also, the `self._process_build()` call at the end of every exit from the context manager looks like there should be a better way to do this.

Have you considered using a single exit stack, use `push()` to add all the Raid/Integrity/LUKS but thereby preserve the previous & much easier to follow code structure?

Reply:
I knew you would say that ;) I also thought while writing it that this is not optimal. However, from the conversation about context managers I had the impression that you wanted this type of nested structure, no ?

I'm fine with an ExitStack that is much better. Will adapt the PR

Reply:
done

Reply:
> However, from the conversation about context managers I had the impression that you wanted this type of nested structure, no ?

No, I actually don't prefer this deep nesting that context managers cause. Hence I am very happy that the exit stack exists, as it allows us to combine the advantages of context managers and not having a five level deep nesting.